yarn-project/end-to-end/src/e2e_persistence.test.ts

```diff
@@ -143,42 +189,158 @@ describe('Aztec persistence', () => {
+    beforeAll(async () => {
+      const temporaryContext = await setup(0, { deployL1ContractsValues }, {});
+
+      await temporaryContext.pxe.addContracts([
+        {
+          artifact: TokenContract.artifact,
+          completeAddress: contractAddress,
+          portalContract: EthAddress.ZERO,
+        },
+      ]);
+
+      const ownerAccount = getUnsafeSchnorrAccount(temporaryContext.pxe, ownerPrivateKey, ownerAddress);
+      await ownerAccount.register();
+      const ownerWallet = await ownerAccount.getWallet();
+
+      const contract = await TokenContract.at(contractAddress.address, ownerWallet);
+
+      // mint some tokens with a secret we know and redeem later on a separate PXE
+      secret = Fr.random();
+      mintAmount = 1000n;
+      const mintTxReceipt = await contract.methods
+        .mint_private(mintAmount, computeMessageSecretHash(secret))
+        .send()
+        .wait();
+      mintTxHash = mintTxReceipt.txHash;
+
+      // publicly reveal that I have 1000 tokens
+      revealedAmount = 1000n;
+      await contract.methods.unshield(ownerAddress, ownerAddress, revealedAmount, 0).send().wait();
+
+      // shut everything down
+      await temporaryContext.teardown();
+    }, 100_000);
+
+    let ownerWallet: AccountWallet;
+    let contract: TokenContract;
+
+    beforeEach(async () => {
+      context = await setup(0, { dataDirectory, deployL1ContractsValues }, { dataDirectory });
+      ownerWallet = await getUnsafeSchnorrWallet(context.pxe, ownerAddress.address, ownerPrivateKey);
+      contract = await TokenContract.at(contractAddress.address, ownerWallet);
+
+      await waitForAccountSynch(context.pxe, ownerAddress, { interval: 0.1, timeout: 5 });
+    }, 5000);
+
+    afterEach(async () => {
+      await context.teardown();
+    });
+
+    it("restores owner's public balance", async () => {
+      await expect(contract.methods.balance_of_public(ownerAddress.address).view()).resolves.toEqual(revealedAmount);
+    });
+
+    it('allows consuming transparent note created on another PXE', async () => {
+      // this was created in the temporary PXE in `beforeAll`
+      await addPendingShieldNoteToPXE(
+        ownerWallet,
+        contractAddress,
+        mintAmount,
+        computeMessageSecretHash(secret),
+        mintTxHash,
+      );
+
+      const balanceBeforeRedeem = await contract.methods.balance_of_private(ownerWallet.getAddress()).view();
+
+      await contract.methods.redeem_shield(ownerWallet.getAddress(), mintAmount, secret).send().wait();
+      const balanceAfterRedeem = await contract.methods.balance_of_private(ownerWallet.getAddress()).view();
+
+      expect(balanceAfterRedeem).toEqual(balanceBeforeRedeem + mintAmount);
     });
   });
 });
+
+async function addPendingShieldNoteToPXE(
+  wallet: AccountWallet,
+  asset: CompleteAddress,
+  amount: bigint,
+  secretHash: Fr,
+  txHash: TxHash,
+) {
+  // The storage slot of `pending_shields` is 5.
+  // TODO AlexG, this feels brittle
```

Comment:
I think it would be cool if we extracted storage slots from the Noir code and made them available on the Contract interface. Something like this would be really nice to use
```ts
async function addPendingShieldNoteToPXE(
  wallet: AccountWallet,
  asset: TokenContract,
  amount: bigint,
  secretHash: Fr,
  txHash: TxHash,
) {
  const note = new Note([new Fr(amount), secretHash]);
  const extendedNote = new ExtendedNote(note, wallet.getAddress(), contract.address, contract.storageSlots.pendingShields, txHash);
  await wallet.addNote(extendedNote);
}
```

Reply:
Related to https://github.com/AztecProtocol/aztec-packages/issues/2806. Having a mapping of a contract storage available has popped up multiple times.